File: submission-tools/model_template.py
"""Starter program: paste your model into ``forecast_model`` below.

Run from the repository root:

    python submission-tools/model_template.py YYYY-MM-DD team-model
    python submission-tools/model_template.py --all-historical team-model

The program selects the correct retrospective input when one exists. For a
live round it reads target-data/time-series.csv. In both cases, observations
after the Sunday preceding reference_date are removed before the model runs.
Only the Python standard library is required; participants may add their own
dependencies inside forecast_model.
"""

from __future__ import annotations

import argparse
import csv
import math
import re
from datetime import date, timedelta
from pathlib import Path
from typing import Any

LOCATIONS = ("SE", "SE-M", "SE-O")
HORIZONS = (0, 1, 2, 3)
TARGET = "weekly incident influenza cases"
OUTPUT_COLUMNS = (
    "reference_date",
    "target",
    "horizon",
    "location",
    "output_type",
    "output_type_id",
    "value",
)
MODEL_PATTERN = re.compile(r"^[A-Za-z0-9_+]+-[A-Za-z0-9_+]+$")
HISTORICAL_START = date.fromisocalendar(2025, 40, 7)
HISTORICAL_END = date.fromisocalendar(2026, 20, 7)
HISTORICAL_MANIFEST = Path("retrospective-data/2025-2026/manifest.csv")


# ---------------------------------------------------------------------------
# REPLACE THIS FUNCTION BODY WITH YOUR MODEL
# ---------------------------------------------------------------------------
def forecast_model(
    data: list[dict[str, str]], reference_date: date
) -> list[dict[str, Any]]:
    """Return one value for every location and horizon to be submitted.

    ``data`` contains the hub target-data columns and has already been cut at
    reference_date minus seven days. This example is a persistence model: it
    repeats the latest available observation for horizons 0--3.
    """
    forecasts: list[dict[str, Any]] = []
    for location in LOCATIONS:
        observed = [
            row
            for row in data
            if row["location"] == location
            and row["status"] == "available"
            and row["value"].strip()
        ]
        if not observed:
            raise ValueError(f"No available observations for {location}")
        latest = max(observed, key=lambda row: row["target_end_date"])
        for horizon in HORIZONS:
            forecasts.append(
                {
                    "location": location,
                    "horizon": horizon,
                    "value": float(latest["value"]),
                }
            )
    return forecasts


# ---------------------------------------------------------------------------
# HUB FILE HANDLING — NORMALLY NO CHANGES ARE NEEDED BELOW THIS LINE
# ---------------------------------------------------------------------------
def default_input_path(reference_date: date) -> Path:
    retrospective = (
        Path("retrospective-data")
        / "2025-2026"
        / "rounds"
        / f"{reference_date.isoformat()}.csv"
    )
    if retrospective.is_file():
        return retrospective
    return Path("target-data/time-series.csv")


def read_input(path: Path, reference_date: date) -> list[dict[str, str]]:
    cutoff = reference_date - timedelta(days=7)
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle)
        required = {"location", "target_end_date", "value", "status"}
        missing = sorted(required - set(reader.fieldnames or ()))
        if missing:
            raise ValueError(f"Input data is missing columns: {missing}")
        rows = []
        for row in reader:
            try:
                target_end_date = date.fromisoformat(row["target_end_date"])
            except ValueError as exc:
                raise ValueError(
                    f"Invalid target_end_date {row['target_end_date']!r}"
                ) from exc
            if target_end_date <= cutoff:
                rows.append(row)
    if not rows:
        raise ValueError(f"No input observations are available through {cutoff}")
    return rows


def validate_forecasts(forecasts: list[dict[str, Any]]) -> list[dict[str, Any]]:
    if not forecasts:
        raise ValueError("The model returned no forecasts")
    normalized: list[dict[str, Any]] = []
    seen: set[tuple[str, int]] = set()
    locations: set[str] = set()
    for number, forecast in enumerate(forecasts, start=1):
        try:
            location = str(forecast["location"])
            horizon = int(forecast["horizon"])
            value = float(forecast["value"])
        except (KeyError, TypeError, ValueError) as exc:
            raise ValueError(
                f"Forecast {number} must contain location, horizon and numeric value"
            ) from exc
        if location not in LOCATIONS:
            raise ValueError(f"Unsupported location {location!r}")
        if horizon not in HORIZONS:
            raise ValueError(f"Unsupported horizon {horizon!r}")
        if not math.isfinite(value) or value < 0:
            raise ValueError("Forecast values must be finite and at least zero")
        key = (location, horizon)
        if key in seen:
            raise ValueError(f"Duplicate forecast for {location}, horizon {horizon}")
        seen.add(key)
        locations.add(location)
        normalized.append(
            {"location": location, "horizon": horizon, "value": value}
        )
    expected = {(location, horizon) for location in locations for horizon in HORIZONS}
    if seen != expected:
        missing = sorted(expected - seen)
        raise ValueError(f"Every submitted location needs horizons 0-3; missing {missing}")
    return sorted(
        normalized,
        key=lambda row: (LOCATIONS.index(row["location"]), row["horizon"]),
    )


def write_submission(
    path: Path,
    reference_date: date,
    forecasts: list[dict[str, Any]],
) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", encoding="utf-8", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=OUTPUT_COLUMNS)
        writer.writeheader()
        for forecast in forecasts:
            writer.writerow(
                {
                    "reference_date": reference_date.isoformat(),
                    "target": TARGET,
                    "horizon": forecast["horizon"],
                    "location": forecast["location"],
                    "output_type": "mean",
                    "output_type_id": "",
                    "value": forecast["value"],
                }
            )


def historical_reference_dates(path: Path = HISTORICAL_MANIFEST) -> list[date]:
    with path.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle)
        if "reference_date" not in (reader.fieldnames or ()):
            raise ValueError("Historical manifest has no reference_date column")
        result = [date.fromisoformat(row["reference_date"]) for row in reader]
    if len(result) != 33 or len(set(result)) != 33:
        raise ValueError("Historical manifest must contain 33 unique rounds")
    if result != sorted(result) or any(value.weekday() != 6 for value in result):
        raise ValueError("Historical manifest dates must be ordered Sundays")
    return result


def prepare_round(
    reference_date: date,
    *,
    input_path: Path,
) -> list[dict[str, Any]]:
    data = read_input(input_path, reference_date)
    forecasts = validate_forecasts(forecast_model(data, reference_date))
    submitted_locations = {forecast["location"] for forecast in forecasts}
    if (
        HISTORICAL_START <= reference_date <= HISTORICAL_END
        and submitted_locations != set(LOCATIONS)
    ):
        raise ValueError("Historical rounds require forecasts for all locations")
    return forecasts


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "arguments",
        nargs="+",
        metavar="ARG",
        help="YYYY-MM-DD and model_id, or only model_id with --all-historical",
    )
    parser.add_argument(
        "--all-historical",
        action="store_true",
        help="Run every round listed in the 2025/2026 manifest",
    )
    parser.add_argument("--input-data", type=Path)
    parser.add_argument("--output", type=Path)
    parser.add_argument("--output-root", type=Path, default=Path("model-output"))
    args = parser.parse_args()

    if args.all_historical:
        if len(args.arguments) != 1:
            parser.error("--all-historical requires exactly one model_id")
        if args.input_data or args.output:
            parser.error("--input-data and --output cannot be used with --all-historical")
        model_id = args.arguments[0]
        try:
            reference_dates = historical_reference_dates()
        except (OSError, ValueError) as exc:
            raise SystemExit(f"Could not read historical manifest: {exc}") from exc
    else:
        if len(args.arguments) != 2:
            parser.error("provide reference_date and model_id")
        reference_date_text, model_id = args.arguments
        try:
            reference_date = date.fromisoformat(reference_date_text)
        except ValueError:
            parser.error("reference_date must use YYYY-MM-DD format")
        if reference_date.weekday() != 6:
            parser.error("reference_date must be a Sunday")
        reference_dates = [reference_date]

    if not MODEL_PATTERN.fullmatch(model_id):
        parser.error("model_id must have the form team-model")

    try:
        prepared = []
        for reference_date in reference_dates:
            input_path = args.input_data or default_input_path(reference_date)
            output_path = args.output or (
                args.output_root
                / model_id
                / f"{reference_date.isoformat()}-{model_id}.csv"
            )
            prepared.append(
                (
                    reference_date,
                    input_path,
                    output_path,
                    prepare_round(reference_date, input_path=input_path),
                )
            )
        for reference_date, input_path, output_path, forecasts in prepared:
            write_submission(output_path, reference_date, forecasts)
            print(f"Read {input_path}")
            print(f"Wrote {output_path}")
    except (OSError, ValueError) as exc:
        raise SystemExit(f"Could not create forecast: {exc}") from exc
    if args.all_historical:
        print(f"Wrote {len(prepared)} historical forecast files")


if __name__ == "__main__":
    main()

File: submission-tools/model_template.R
# Starter program: replace only the body of forecast_model with your model.
# Run from the repository root:
#
#   Rscript submission-tools/model_template.R YYYY-MM-DD team-model
#   Rscript submission-tools/model_template.R --all-historical team-model
#
# Add an optional input CSV and output CSV as the third and fourth arguments.

locations <- c("SE", "SE-M", "SE-O")
horizons <- 0:3
target_name <- "weekly incident influenza cases"

# ---------------------------------------------------------------------------
# REPLACE THIS FUNCTION BODY WITH YOUR MODEL
# ---------------------------------------------------------------------------
forecast_model <- function(data, reference_date) {
  # data has already been cut at reference_date minus seven days.
  # This example is a persistence model that repeats the latest observation.
  output <- do.call(rbind, lapply(locations, function(location) {
    observed <- data[
      data$location == location &
        data$status == "available" &
        !is.na(data$value),
      ,
      drop = FALSE
    ]
    if (nrow(observed) == 0) {
      stop("No available observations for ", location)
    }
    latest <- observed[which.max(as.Date(observed$target_end_date)), ]
    data.frame(
      location = location,
      horizon = horizons,
      value = rep(as.numeric(latest$value), length(horizons)),
      stringsAsFactors = FALSE
    )
  }))
  output
}

# ---------------------------------------------------------------------------
# HUB FILE HANDLING — NORMALLY NO CHANGES ARE NEEDED BELOW THIS LINE
# ---------------------------------------------------------------------------
default_input_path <- function(reference_date) {
  retrospective <- file.path(
    "retrospective-data", "2025-2026", "rounds",
    paste0(reference_date, ".csv")
  )
  if (file.exists(retrospective)) {
    retrospective
  } else {
    file.path("target-data", "time-series.csv")
  }
}

prepare_round <- function(reference_date, model_id, input_path, output_path) {
  data <- read.csv(
    input_path,
    stringsAsFactors = FALSE,
    check.names = FALSE,
    na.strings = c("", "NA", "N/A")
  )
  required_input <- c("location", "target_end_date", "value", "status")
  missing_input <- setdiff(required_input, names(data))
  if (length(missing_input) > 0) {
    stop("Input data is missing columns: ", paste(missing_input, collapse = ", "))
  }
  target_dates <- as.Date(data$target_end_date)
  if (any(is.na(target_dates))) {
    stop("Input data contains an invalid target_end_date")
  }
  cutoff <- reference_date - 7
  data <- data[target_dates <= cutoff, , drop = FALSE]
  if (nrow(data) == 0) {
    stop("No input observations are available through ", cutoff)
  }

  forecasts <- forecast_model(data, reference_date)
  required_forecast <- c("location", "horizon", "value")
  if (!is.data.frame(forecasts) || !all(required_forecast %in% names(forecasts))) {
    stop("forecast_model must return a data frame with location, horizon and value")
  }
  if (nrow(forecasts) == 0) {
    stop("forecast_model returned no forecasts")
  }
  if (any(!forecasts$location %in% locations)) {
    stop("forecast_model returned an unsupported location")
  }
  forecast_horizons <- suppressWarnings(as.integer(forecasts$horizon))
  forecast_values <- suppressWarnings(as.numeric(forecasts$value))
  if (any(is.na(forecast_horizons)) || any(!forecast_horizons %in% horizons)) {
    stop("forecast_model returned an unsupported horizon")
  }
  if (any(is.na(forecast_values)) || any(!is.finite(forecast_values)) || any(forecast_values < 0)) {
    stop("Forecast values must be finite and at least zero")
  }
  keys <- paste(forecasts$location, forecast_horizons, sep = ":")
  if (anyDuplicated(keys)) {
    stop("forecast_model returned a duplicate location and horizon")
  }
  for (location in unique(forecasts$location)) {
    found <- sort(forecast_horizons[forecasts$location == location])
    if (!identical(found, horizons)) {
      stop("Every submitted location needs horizons 0-3")
    }
  }
  historical_start <- as.Date("2025-10-05")
  historical_end <- as.Date("2026-05-17")
  if (
    reference_date >= historical_start &&
      reference_date <= historical_end &&
      !setequal(unique(forecasts$location), locations)
  ) {
    stop("Historical rounds require forecasts for all locations")
  }

  submission <- data.frame(
    reference_date = rep(as.character(reference_date), nrow(forecasts)),
    target = rep(target_name, nrow(forecasts)),
    horizon = forecast_horizons,
    location = forecasts$location,
    output_type = rep("mean", nrow(forecasts)),
    output_type_id = rep(NA_character_, nrow(forecasts)),
    value = forecast_values,
    stringsAsFactors = FALSE
  )
  submission$location <- factor(submission$location, levels = locations)
  submission <- submission[order(submission$location, submission$horizon), ]
  submission$location <- as.character(submission$location)
  list(
    reference_date = reference_date,
    input_path = input_path,
    output_path = output_path,
    submission = submission
  )
}

write_round <- function(prepared) {
  dir.create(dirname(prepared$output_path), recursive = TRUE, showWarnings = FALSE)
  write.csv(prepared$submission, prepared$output_path, row.names = FALSE, na = "")
  message("Read ", prepared$input_path)
  message("Wrote ", prepared$output_path)
}

args <- commandArgs(trailingOnly = TRUE)
usage <- paste(
  "Usage:",
  "Rscript submission-tools/model_template.R YYYY-MM-DD team-model [input.csv] [output.csv]",
  "or Rscript submission-tools/model_template.R --all-historical team-model [output-root]",
  sep = "\n  "
)
if (length(args) < 2) {
  stop(usage)
}

all_historical <- identical(args[[1]], "--all-historical")
model_id <- args[[2]]
if (!grepl("^[A-Za-z0-9_+]+-[A-Za-z0-9_+]+$", model_id)) {
  stop("model_id must have the form team-model")
}

if (all_historical) {
  if (length(args) > 3) {
    stop(usage)
  }
  output_root <- if (length(args) == 3) args[[3]] else "model-output"
  manifest_path <- file.path("retrospective-data", "2025-2026", "manifest.csv")
  manifest <- read.csv(manifest_path, stringsAsFactors = FALSE)
  reference_dates <- as.Date(manifest$reference_date)
  if (
    length(reference_dates) != 33 ||
      length(unique(reference_dates)) != 33 ||
      any(is.na(reference_dates)) ||
      !identical(reference_dates, sort(reference_dates)) ||
      any(as.POSIXlt(reference_dates)$wday != 0)
  ) {
    stop("Historical manifest must contain 33 unique ordered Sundays")
  }
  prepared <- lapply(reference_dates, function(reference_date) {
    input_path <- default_input_path(reference_date)
    output_path <- file.path(
      output_root, model_id, paste0(reference_date, "-", model_id, ".csv")
    )
    prepare_round(reference_date, model_id, input_path, output_path)
  })
  invisible(lapply(prepared, write_round))
  message("Wrote ", length(prepared), " historical forecast files")
} else {
  reference_date <- as.Date(args[[1]])
  if (is.na(reference_date) || as.POSIXlt(reference_date)$wday != 0) {
    stop("reference_date must be a Sunday in YYYY-MM-DD format")
  }
  input_path <- if (length(args) >= 3) args[[3]] else default_input_path(reference_date)
  output_path <- if (length(args) >= 4) args[[4]] else file.path(
    "model-output", model_id, paste0(reference_date, "-", model_id, ".csv")
  )
  if (length(args) > 4) {
    stop(usage)
  }
  write_round(prepare_round(reference_date, model_id, input_path, output_path))
}
